search-books/src/main/java/nix/education/service/TransactionalDataService.java

```diff
@@ -0,0 +1,38 @@
+package nix.education.service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import nix.education.entity.Book;
+import nix.education.entity.Tag;
+import nix.education.repository.BookRepository;
+import nix.education.repository.TagRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+public class TransactionalDataService {
+
+    public List<String> getTagTitles(Book book) {
+        return book.getTags().stream().map(Tag::getTag).collect(Collectors.toList());
+    }
+
+    public List<Book> getBooksByIds(List<Long> ids, BookRepository bookRepository) {
+        List<Book> books = new ArrayList<>();
+        ids.forEach(id -> books.add(bookRepository.findById(id).orElseThrow()));
+        return books;
+    }
+
+    public List<Tag> getResolvedTags(Book book, TagRepository tagRepository) {
+        List<Tag> tags = new ArrayList<>();
+
+        for (Tag tag : book.getTags()) {
+            if (!tagRepository.existsTagByTag(tag.getTag())) {
+                Tag newTag = new Tag(tag.getTag());
+                tags.add(newTag);
+            } else {
+                tags.add(tagRepository.findTagByTag(tag.getTag()));
+            }
+        }
+        return tags;
+    }
```

Comment:
May be we can try do refactor the code in this way:
```java
private void setTags(Book book) {
        List<Tag> bookTags = book.getTags();
        List<String> tagValues = getTagsList(book);
        List<Tag> existingTags = tagRepository.findByTagIn(tagValues);
        bookTags.removeAll(existingTags);

        List<Tag> resolvedTags = new ArrayList<>();
        resolvedTags.addAll(existingTags);
        resolvedTags.addAll(bookTags);

        book.setTags(resolvedTags);
        book.setTagValues(tagValues);
    }
```

Reply:
Changed logic for searching existed tags. But this variant was refactored a little bit.